### PR TITLE
PolicyRepository: index and replace rules by resource.

### PR DESCRIPTION
**Pull request by squeed**
(Note for reviewers: this also contains a commit fixing some improper use of shared state in tests; it is mostly mechanical but required for this change).

The PolicyRepository is a database of all known network policies (KNP / CNP / CCNP / gRPC Policies). It references policies by Labels, which are an arbitrary set of identifiers for a policy. When an "upstream" policy is updated, a synthetic set of labels is created and used as a key for replacement. 

This label-oriented referencing is inefficient. When a policy is upserted, all existing rules must be scanned to se if they are candidates for replacement. Furthermore, this doesn't reflect how policy rules are actually created. Namely, they have an upstream owning resource, which creates one or more downstream rules in the repository.

This change reorients the PolicyRepository to index and replace rules on a per-resource basis. This pattern is already well established within Cilium, most notably the IPCache, and has proven itself. It also means that the standard policy actions do not require a whole-repository scan or evaluating labels.

The existing label-based replacement mechanism must be preserved for policies managed by the local API, so the existing code cannot be entirely removed, but it will seldom be used in the field.

(This PR was inspired by #27163, but takes a different tack).

```release-note
Starting cilium-agent with large numbers of network policies should be much faster.
```

### CODE REVIEW

**Comment by squeed** [comment]
Tagging @aanm for review too; you understand this bit of code somewhat.

File: pkg/policy/fuzz_test.go [hunk 1]
```diff
@@ -33,7 +33,8 @@ func FuzzResolveEgressPolicy(f *testing.F) {
 		}
 		rule := &rule{Rule: r}
 		state := traceState{}
-		_, _ = rule.resolveEgressPolicy(testPolicyContext, fromBar, &state, L4PolicyMap{}, nil, nil)
+		td := newTestData()
```
Comment (doniacld) [inline pkg/policy/fuzz_test.go:36]: nit: Typo in the commit message `identities`

File: pkg/policy/l4_filter_test.go [hunk 2]
```diff
@@ -15,52 +15,88 @@ import (
...
+
+// resetRepo clears only the policy repository.
+// Some tests rely on the accumulated state, but a clean repo.
+func (td *testData) resetRepo() *Repository {
+	td.repo = NewPolicyRepository(nil, nil, nil, nil)
+	td.repo.selectorCache = td.sc
+	return td.repo
+}
+
+// the testPolicyContext is a
```
Comment (doniacld) [inline pkg/policy/l4_filter_test.go]: Did you not finish your sentence in purpose?

File: pkg/policy/repository.go [hunk 3]
```diff
@@ -397,20 +424,86 @@ func (p *Repository) Add(r api.Rule) (uint64, map[uint16]struct{}, error) {
...
+
+// ReplaceByResourceLocked replaces all rules that belong to a given resource with a
+// new set. The set of rules added and removed is returned, along with the new revision number.
+// Resource must not be empty
+func (p *Repository) ReplaceByResourceLocked(rules api.Rules, resource ipcacheTypes.ResourceID) (newRules ruleSlice, oldRules ruleSlice, revision uint64) {
+	if len(resource) == 0 {
+		// This should never ever be hit, as the caller should have already validated the resource.
+		// However, if it does happen, it means something very wrong has happened and we are at risk
+		// of removing all network policies. So, we must panic rather than risk disabling network security.
+		panic("may not replace API rules with an empty resource")
```
Comment (doniacld) [inline pkg/policy/repository.go:450]: Don't we have a more polite way to return an error or alert?

File: pkg/policy/repository_test.go [hunk 4]
```diff
@@ -300,8 +301,8 @@ func TestAddSearchDelete(t *testing.T) {
 
 	// search rule1,rule2
 	repo.Mutex.RLock()
-	require.EqualValues(t, api.Rules{&rule1, &rule2}, repo.SearchRLocked(lbls1))
-	require.EqualValues(t, api.Rules{&rule3}, repo.SearchRLocked(lbls2))
+	assert.ElementsMatch(t, api.Rules{&rule1, &rule2}, repo.SearchRLocked(lbls1))
```
Comment (doniacld) [inline pkg/policy/repository_test.go]: If something goes wrong, the test will continue and will try to remove by label even if rules are not present. Do you want to keep assert?

File: pkg/policy/rule.go [hunk 5]
```diff
@@ -11,16 +11,22 @@ import (
 	"github.com/cilium/proxy/pkg/policy/api/kafka"
 
 	"github.com/cilium/cilium/pkg/identity"
+	ipcacheTypes "github.com/cilium/cilium/pkg/ipcache/types"
```
Comment (doniacld) [inline pkg/policy/rule.go]: nit:
```
ipcachetypes "github.com/cilium/cilium/pkg/ipcache/types"
```

**Review by christarazi — APPROVED** [review]
LGTM, very nice cleanup! Do we have any performance benchmark numbers that can tell us how much this improves the startup time?

File: pkg/policy/repository_test.go (hunk 4 above)
Comment (squeed) [reply pkg/policy/repository_test.go]: oh, good point, I was debugging something but forgot to switch them back

File: pkg/policy/repository.go (hunk 3 above)
Comment (squeed) [reply pkg/policy/repository.go:450]: We will intercept this earlier on; this is a last-gasp effort. Normally we could ignore the policy but in this case, I don't want to drop policies on the floor.

File: pkg/policy/rule.go (hunk 5 above)
Comment (squeed) [reply pkg/policy/rule.go]: fixed.

**Comment by squeed** [comment]
> Do we have any performance benchmark numbers that can tell us how much this improves the startup time?

I threw together a quick benchmark:

```
$ benchstat bench-before.txt bench-after.txt 
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/policy
cpu: 12th Gen Intel(R) Core(TM) i7-1250U
            │ bench-before.txt │           bench-after.txt            │
            │      sec/op      │    sec/op     vs base                │
AddRules-12      241.740µ ± 7%   1.538µ ± 49%  -99.36% (p=0.000 n=10)
```

Good enough for me  :-)

**Comment by squeed** [comment]
Tests caught a flake -- just some test code that expected a certain ordering. Fixed.

**Review by doniacld — APPROVED** [review]

File: pkg/policy/rule.go [hunk 6]
```diff
@@ -11,16 +11,22 @@ import (
...
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/policy/api"
 )
 
+type ruleKey struct {
+	resource ipcachetypes.ResourceID
+	idx      uint
```
Comment (aanm) [inline pkg/policy/rule.go:24]: Index in reference to what? A comment here would be good.

**Review by aanm — CHANGES_REQUESTED** [review]
@squeed I'm 15 mins too late but I left a comment that I believe it should be addressed.

**Comment by aanm** [comment]
> > Do we have any performance benchmark numbers that can tell us how much this improves the startup time?
> 
> @christarazi I threw together a quick benchmark:
> 
> ```
> $ benchstat bench-before.txt bench-after.txt 
> goos: linux
> goarch: amd64
> pkg: github.com/cilium/cilium/pkg/policy
> cpu: 12th Gen Intel(R) Core(TM) i7-1250U
>             │ bench-before.txt │           bench-after.txt            │
>             │      sec/op      │    sec/op     vs base                │
> AddRules-12      241.740µ ± 7%   1.538µ ± 49%  -99.36% (p=0.000 n=10)
> ```
> 
> Good enough for me :-)

I also ran a local benchmark to compare the PR against current main and I didn't see any regressions.

File: pkg/policy/rule.go (hunk 6 above)
Comment (squeed) [reply pkg/policy/rule.go:24]: sounds good, I'll add a comment in an incoming PR.

**Pull request merged**